File: pari3g.c
#include <stdio.h>
#include <stdlib.h>
struct Node {
    int data;
    struct Node* next;
};

void printList(struct Node* head); 

void removeLoop(struct Node* head) {
    if (head == NULL || head->next == NULL) return;

    struct Node *slow = head, *fast = head;

    slow = slow->next;
    fast = fast->next->next;
    while (fast && fast->next) {
        if (slow == fast) break;
        slow = slow->next;
        fast = fast->next->next;
    }

    if (slow == fast) {
        slow = head;
        if (slow == fast) {
             while(fast->next != slow) fast = fast->next;
        } else {
            while (slow->next != fast->next) {
                slow = slow->next;
                fast = fast->next;
            }
        }
        fast->next = NULL;
    }
}
int main() {
    struct Node* head = (struct Node*)malloc(sizeof(struct Node));
    head->data = 1;
    head->next = (struct Node*)malloc(sizeof(struct Node));
    head->next->data = 2;
    head->next->next = (struct Node*)malloc(sizeof(struct Node));
    head->next->next->data = 3;
    head->next->next->next = head->next; 

    printf("List has a loop. Removing it...\n");
    removeLoop(head);
    printf("List after removing loop: ");
    printList(head);

    return 0;
}
void printList(struct Node* head) { while (head) { printf("%d -> ", head->data); head = head.next; } printf("NULL\n"); }

File: pari3g.py
def remove_loop(self):
    if self.head is None or self.head.next is None:
        return
        
    slow_ptr = self.head
    fast_ptr = self.head

    while fast_ptr and fast_ptr.next:
        slow_ptr = slow_ptr.next
        fast_ptr = fast_ptr.next.next
        if slow_ptr == fast_ptr:
            break

    if slow_ptr != fast_ptr:
        return

    slow_ptr = self.head
    while slow_ptr.next != fast_ptr.next:
        slow_ptr = slow_ptr.next
        fast_ptr = fast_ptr.next

    fast_ptr.next = None

llist = SinglyLinkedList()
llist.insert_at_end(1)
llist.insert_at_end(2)
llist.insert_at_end(3)
llist.insert_at_end(4)
llist.insert_at_end(5)

llist.head.next.next.next.next.next = llist.head.next.next
print("Loop detected:", llist.detect_loop())

llist.remove_loop()
print("Loop detected after removal:", llist.detect_loop())

print("List after removing loop:")
llist.print_list()
